src/core/features/tags/controller/TagsController.ts

```diff
@@ -55,24 +99,100 @@ export class TagsController {
 	}
 
 	public async add(name: string, parent: null | string): Promise<string> {
-		const db = wrapDB(this.db.get());
-
-		const segments = name.split('/');
+		validateTagName(name);
 
 		let lastId: string | null = null;
-		if (segments.length > 1) {
-			await this.db.get().transaction(async (tx) => {
-				const db = wrapDB(tx);
 
-				for (let idx = 0; idx < segments.length; idx++) {
-					const name = segments[idx];
+		await this.db.get().transaction(async (tx) => {
+			const db = wrapDB(tx);
+
+			let resolvedTagToCreate: string;
+			if (!parent) {
+				resolvedTagToCreate = name;
+			} else {
+				const {
+					rows: [parentResolvedTag],
+				} = await db.query(
+					qb.sql`SELECT resolved_name FROM (${selectResolvedTags(
+						this.workspace,
+					)}) WHERE id=${parent} 
+					LIMIT 1`,
+					resolvedNameSchema,
+				);
+				// If parent tag is not found in the database, the tag cannot be created
+				if (parent && !parentResolvedTag)
+					throw new TagControllerError(
+						`Parent tag: ${parent} not exist`,
+						TAG_ERROR_CODE.TAG_NOT_EXIST,
+					);
+				resolvedTagToCreate = `${parentResolvedTag.resolvedName}/${name}`;
+			}
+
+			// Check tag uniqueness
+			const {
+				rows: [duplicatedTag],
+			} = await db.query(
+				qb.sql`SELECT resolved_name FROM (${selectResolvedTags(
+					this.workspace,
+				)}) WHERE resolved_name = ${resolvedTagToCreate} 
+					LIMIT 1`,
+				resolvedNameSchema,
+			);
+			if (duplicatedTag) {
+				throw new TagControllerError(
+					`Tag ${duplicatedTag.resolvedName} already exists`,
+					TAG_ERROR_CODE.DUPLICATE,
+				);
+			}
+
+			const segments = resolvedTagToCreate.split('/');
+			if (segments.length > 1) {
+				// Create an array of names variants from a string: 'foo/bar' - ['foo', 'foo/bar']
```

Comment:
Explain the purpose of this operation. As i understand the idea is to pick only those resolved names that match exactly with a path segment, not partially.

For example, in case we have tag `/foo/bar` and user request to create `/foo/bad/baz` it will found only `/foo`, since `bar` match only 2 of 3 chars